### DISCUSSION TITLE
Using Python's Typing module for NumPy types

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1489] [Ideas]
In a comment on a PR we discussed type specification:

> > >               Can we please stop using the `types` decorator in our tests, and use type annotations instead?
> > > 
> > > @EmilyBourne I think that we should deprecate this decorator, which makes the code not compatible with pure Python.
> > > 
> > > _Originally posted by @yguclu in https://github.com/pyccel/pyccel/pull/1280#discussion_r1277562709_
> > 
> >
> >               This has been discussed multiple times. It is not incompatible with pure python if the import is correctly included `from pyccel.decorators import types`. It can be replaced by `@template` to some degree, but it gives more flexibility to limit the number of cases generated. E.g:
> > ```
> > @types('float[:]','int[:]')
> > @types('float[:,:]','int[:,:]')
> > @types('float[:,:,:]','int[:,:,:]')
> > def f(a, b):
> >     pass
> > ```
> > which describes 3 cases where arrays have the same number of dimensions. The equivalent with templates would lead to 9 cases including cases where arrays have different numbers of dimensions which could potentially be problematic:
> > ```
> > @template('T', ['float[:]','float[:,:]','float[:,:,:]'])
> > @template('T', ['int[:]','int[:,:]','int[:,:,:]'])
> > def f(a, b):
> >     pass
> > ```
> > 
> > _Originally posted by @EmilyBourne in https://github.com/pyccel/pyccel/pull/1280#discussion_r1277613335_
> >             
> 
> True, there are some use cases for the @types annotation, but in most instances they could be replaced with either the @elemental annotation or the "partial templates", once we get to finally implement them.
> _Originally posted by @yguclu in https://github.com/pyccel/pyccel/pull/1280#discussion_r1278411358_

Assuming that we want to prefer the pure Python implementation where possible we could also extend this to NumPy functions.

Currently we have syntax such as:
```python

def f(a : 'float[:]', b : 'int[:]'):
    pass
```

without pyccel decorators this would be written as:
```python
import numpy.typing as npt
def f(a : npt.NDArray[float], b : npt.NDArray[int]):
    pass
```

See https://numpy.org/devdocs/reference/typing.html

**Comment by yguclu** [comment]
@EmilyBourne In your example does NumPy typing specify that the arrays are 1D? Pyccel needs that information.

**Reply by EmilyBourne** [reply]
I assumed it would and just needed digging into but I have just found this:

> Until the introduction of shape typing (see [PEP 646](https://peps.python.org/pep-0646/)) it is unfortunately not possible to make the necessary distinction between 0D and >0D arrays.

But this seems to refer to NumPy's code rather than code in general. I think we would have to dig through their mypy code to confirm this as their documentation is not yet as detailed as it could be

**Reply by adam-urbanczyk** [reply]
Note that similar projects (numba and lpython) define their own types (e.g.  `float64`) and use those to annotate arrays (e.g. `float64[:,:,:]`)
